File: contestants/alice.c
/* this module can be used for testing your solution */

#include <stdio.h>
#include <stdlib.h>
#include "alice.h"

static int n = 0;

static int l, questions_number;
	
int get_n()
{
  if (n == 0) {
    printf("Input n - the upper bound on the number Alice has to choose: ");
    scanf("%d", &n);
    if (n <= 0) {
      printf("ERROR: n nonpositive\n");
      exit(1);
    }
    printf("Input Alice's secret number: ");
    scanf("%d", &l);
    if (l <= 0 || l > n) {
      printf("ERROR: Alice's number out of bounds\n");
      exit(1);
    }
    questions_number = 0;
  }
  printf("get_n() = %d\n", n);
  return n;
}

int is_divisible_by(int m)
{
  int ans;
  printf("is_divisible_by(%d) ", m);
  if (n == 0) {
    printf("ERROR: "
	   "Function `is_divisible_by' was called before function `get_n'\n");
    exit(1);
  }
  if (m <= 0) {
    printf("ERROR: Function `is_divisible_by' was called with negative m\n");
    exit(1);
  }
  if (m > n) {
    printf("ERROR: Function `is_divisible_by' was called with m > n\n");
    exit(1);
  }
  questions_number++;
  ans = l % m == 0;
  printf("= %d\n", ans);
  return ans;
}
	
void guess(int k)
{
  printf("guess(%d)", k);
  if (n == 0) {
    printf(" ERROR: Function `guess' was called before function `get_n'\n");
    exit(1);
  }

  printf("\nAlice's secret number is %d\n", l);

  if (k != l) {
    printf("FAIL: Alice's secret number is missed after asking %d questions\n",
	   questions_number);
    exit(1);
  }

  printf("SUCCESS: Alice's secret number guessed successively after asking "
	 "%d questions\n", questions_number);
  exit(0);
}

File: contestants/alice.pas
unit alice;

{ this module can be used for testing your solution }

interface

function get_n: longint;
{ initialize the game,
 returns n - the upper bound on the number
 that Alice has in mind }

function is_divisible_by(m : longint) : boolean;
{ question whether m divides Alice's number }
{ input: m > 0 }

procedure guess(k : longint);
{ attempts to guess Alice's number;
 should be called only once }

implementation

var l, n, questions_number: longint;

function get_n: longint;
begin
   if n = 0 then begin
      write('Input n - the upper bound on the number Alice has to choose: ');
      readln(n);
      if n <= 0 then begin
	 writeln('ERROR: n nonpositive');
	 halt(1);
      end;
      write('Input Alice''s secret number: ');
      readln(l);
      if (l <= 0) or (l > n) then begin
	 writeln('ERROR: Alice''s number out of bounds');
	 halt(1);
      end;
      questions_number := 0;
   end;
   writeln('get_n() = ', n);
   get_n := n;
end;

function is_divisible_by(m : longint) : boolean;
begin
   write('is_divisible_by(', m, ') ');
   if n = 0 then begin
      writeln('ERROR: Function `is_divisible_by'' was called before function `get_n''');
      halt(1);
   end;
   if m <= 0 then begin
      writeln('ERROR: Function `is_divisible_by'' was called with negative m');
      halt(1);
   end;
   if m > n then begin
      writeln('ERROR: Function `is_divisible_by'' was called with m > n');
      halt(1);
   end;
   inc(questions_number);
   is_divisible_by := l mod m = 0;
   writeln('= ', is_divisible_by);
end;

procedure guess(k : longint);
begin
   write('guess(', k, ')');

   if n = 0 then begin
      writeln(' ERROR: Procedure `guess'' was called before function `get_n''');
      halt(1);
   end;

   writeln;
   writeln('Alice''s secret number is ', l);

   if k <> l then begin
      writeln('FAIL: Alice''s secret number is missed after asking ',
	      questions_number, ' questions');
      halt(1);
   end;

   writeln('SUCCESS: Alice''s secret number guessed successively after asking ',
	   questions_number, ' questions');
   halt;
end;	

begin
   n := 0;
end.
